python/pyarrow/array.pxi

```diff
@@ -241,6 +246,13 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
 
     if hasattr(obj, '__arrow_array__'):
         return _handle_arrow_array_protocol(obj, type, mask, size)
+    elif hasattr(obj, '__arrow_c_array__'):
+        if type is not None:
+            requested_type = type.__arrow_c_schema__()
```

Comment:
If this raises a `AttributeError`, we should perhaps catch it and raise a `TypeError` with a more descriptive error message? The average user would not know what a `__arrow_c_schema__` attribute implies.

Reply:
Ah, thanks for the explanation.